File: 238 Product of Array Except Self/238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        # Initialize result list with 1s, same length as input
        total = [1] * n

        # Initialize left product accumulator
        left = 1
        # First pass: Left to right
        for i in range(n):
            # Store the product of all elements to the left
            total[i] = left
            # Update left product for next iteration
            left *= nums[i]
        
        # Initialize right product accumulator
        right = 1
        # Second pass: Right to left
        for i in range(n-1, -1, -1):
            # Multiply stored left product with right product
            total[i] *= right
            # Update right product for next iteration
            right *= nums[i]
        
        # Return the result list
        return total

File: arrays/238 Product of Array Except Self/238.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        // Initialize result vector with 1s, same size as input
        vector<int> total(n, 1);

        // Initialize left and right product accumulators
        int left = 1;
        int right = 1;

        // First pass: Left to right
        for (int i = 0; i < n; i++) {
            // Store the product of all elements to the left
            total[i] = left;
            // Update left product for next iteration
            left *= nums[i];
        }

        // Second pass: Right to left
        for (int i = n - 1; i >= 0; i--) {
            // Multiply stored left product with right product
            total[i] *= right;
            // Update right product for next iteration
            right *= nums[i];
        }

        // Return the result vector
        return total;
    }
};
